subsys/net/lib/lwm2m/lwm2m_obj_firmware.c

```diff
@@ -83,7 +158,16 @@ static int package_uri_write_cb(u16_t obj_inst_id,
 {
 	SYS_LOG_DBG("PACKAGE_URI WRITE: %s", package_uri);
 #ifdef CONFIG_LWM2M_FIRMWARE_UPDATE_PULL_SUPPORT
-	lwm2m_firmware_start_transfer(package_uri);
+	u8_t state = lwm2m_firmware_get_update_state();
+	if (STATE_IDLE == state) {
+		lwm2m_firmware_set_update_result(RESULT_DEFAULT);
+		lwm2m_firmware_start_transfer(package_uri);
+	} else if (STATE_DOWNLOADED == state) {
+		if (data_len == 0) {
```

Comment:
From [LwM2M specification](https://goo.gl/5HYazk) figure 29, it looks like it only reset to state idle when it's currently at state downloaded and an empty package URI/resource is written.
That's why I put the state == STATE_DOWNLOADED check here.

If we reset the state machine at any state, then we have to do (1) cancel the firmware download and (2) cancel the firmware update.

I'm not sure which is the correct behavior. But from my point of view, it's simpler if we only accept the state machine reset at DOWNLOADED state.